### DISCUSSION TITLE
Final position was outside tolerance. Check robot safety limits that could be inhibiting motion (Same issue as #233)

### DISCUSSION THREAD

**Discussion by SergeAlhalbi** [Yaskawa-Global/motoros2#308] [Troubleshooting] [answered]
Hello everyone,

I’m encountering an issue where the robot moves successfully to the first waypoint (in Cartesian space) using the MoveIt Task Constructor (MTC), but then the following error occurs:

```
[move_group-4] [WARN] [1725660470.550437309] [moveit.simple_controller_manager.follow_joint_trajectory_controller_handle]: Controller 'follow_joint_trajectory' failed with error: unknown error: Final position was outside tolerance. Check robot safety limits that could be inhibiting motion. [group_1/joint_5: 0.0000 deviation]
[move_group-4] [WARN] [1725660470.550522354] [moveit_ros.trajectory_execution_manager]: Controller handle follow_joint_trajectory reports status ABORTED
[move_group-4] [INFO] [1725660470.550548538] [moveit_ros.trajectory_execution_manager]: Completed trajectory execution with status ABORTED ...
[task_constructor_example-6] [ERROR] [1725660470.552712298] [moveit_task_constructor_executor_110906006630656]: Goal was aborted or canceled
[task_constructor_example-6] [ERROR] [1725660470.562594942] [mtc_node]: Task execution failed
```

After reaching the first waypoint, I think the error is stopping the entire process, preventing the robot from continuing through the remaining waypoints to generate a complete trajectory.

**Here are some relevant details:**
- OS: Ubuntu 22.04.4
- ROS version: ROS2 Humble
- Robot: Yaskawa GP4
- Controller: YRC1000micro
- MoveIt version: MoveIt 2
- Framework: MoveIt Task Constructor (MTC)

**What I’ve tried:**
- Lowered the tolerance in the ros2_controllers.yaml file, but the error persisted. I also tried including the settings in the moveit_controllers.yaml file (check the attached Zip folder).
- Tried to disable the error check in the ActionServer_FJT.c file, but I couldn’t locate the file. I couldn't find any files or folders in the motoros2 repository; I don't even have the folder. After following the instructions on the motoros2 GitHub page, it wasn’t clear whether I needed to download or clone the main motoros2 repository, as it didn’t explicitly state that.

**Here are my questions:**
- Is this issue specific to Yaskawa robots?
- Has this been resolved?
- Is my method for defining the tolerance correct?
- How can I locate the ActionServer_FJT.c script to turn off the check?

Any help would be greatly appreciated.

Thank you,
Serge

[YAMLS.zip](https://github.com/user-attachments/files/16966956/YAMLS.zip)

**Comment by gavanderhoorn** [comment]
Please provide us with a [debug log](https://github.com/Yaskawa-Global/motoros2/blob/7bac0079de3ce43288bc340e1d3b08a6e69d8fa2/doc/troubleshooting.md#debug-log-client).

Make sure to start the debug logger *before* you start the Yaskawa controller.

A quick comment: `ros_control` is not used by MotoROS2, so changing `ros2_controllers.yaml` won't do anything.

re: changing `ActionServer_FJT.c`: where did you read you should 'disable the check'? That file is part of MotoROS2 and to change it you'd need to build MotoROS2 yourself from source, using the MotoPlus SDK. That's not something we recommend users do, and disabling checks like this is not what we would suggest.

**Reply by SergeAlhalbi** [reply]
Thank you for your response.

- I couldn't find the tools folder to run the debug log. As I mentioned, it seems like I don't have any of the folders or files from the MotoROS2 repository including the MotoROS2 folder itself. Should I clone it? Apologies, I'm still new to these concepts.

- Here is where I read to "disable the check"
https://github.com/Yaskawa-Global/motoros2/issues/233
"@ted-miller: my suggestion would be to disable the final goal_tolerance check for now, provide @Illumo-Vincent / @rsaintobert and @TE-2 with a fixed build and in parallel work on correcting the implementation."

- May I ask where I usually change the tolerance and how to do it? Is there any documentation for the syntax?

**Reply by gavanderhoorn** [reply]
> Thank you for your response.
> 
> * I couldn't find the tools folder to run the debug log. [..] Should I clone it? Apologies, I'm still new to these concepts.

The debug logger is part of the source repository, yes. You can either clone the repository, or simply [download it](https://github.com/Yaskawa-Global/motoros2/archive/refs/heads/main.zip) ([documentation](https://docs.github.com/en/repositories/working-with-files/using-files/downloading-source-code-archives#downloading-source-code-archives)).

> * As I mentioned, it seems like I don't have any of the folders or files from the MotoROS2 repository including the MotoROS2 folder itself.

binary releases (which I believe you downloaded) don't include the source code.

As it's not normally needed we don't add it to the release archives, as it would only confuse users.

> * Here is where I read to "disable the check"
>   [persistent "Final position was outside tolerance" #233](https://github.com/Yaskawa-Global/motoros2/issues/233)
>   "@ted-miller: my suggestion would be to disable the final goal_tolerance check for now, provide @Illumo-Vincent / @rsaintobert and @TE-2 with a fixed build and in parallel work on correcting the implementation."

this is us maintainers talking to each other. I'm specifically addressing @ted-miller with that comment by @-mentioning him ([documentation](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#mentioning-people-and-teams)).

There are no released MotoROS2 builds available which have this feature disabled, or allow users to disable it.

> * May I ask where I usually change the tolerance and how to do it? Is there any documentation for the syntax?

The tolerance is either set by whatever tool you use to generate the `FollowJointTrajectoryAction` goal, or by you manually as part of creating the goal.

(edit: only if no tolerances are configured will MotoROS2 use a [default value](https://github.com/Yaskawa-Global/motoros2/blob/7bac0079de3ce43288bc340e1d3b08a6e69d8fa2/src/ActionServer_FJT.h#L14). We can see whether this is the case in a debug log)

This is not part of MotoROS2. It's part of the goal (ie: `.action`) definition. As such, there is no documentation we can provide you.

Having written that: please still provide us with a debug log. Without it we can't assist you, and it looks like there is still something not entirely correct with how MotoROS2 either parses or checks against goal tolerances.

The preferred course of action would be to diagnose what's going on, not to disable the misbehaving functionality.

**Reply by gavanderhoorn** [reply]
Please also provide us with a copy of the `PANELBOX.LOG` file from your controller.

**Comment by SergeAlhalbi** [comment]
Thank you for the explanation. I just cloned the MotoROS2 repository.

Below are the debug log and the PANELBOX.LOG.

I couldn't find a way to define a tolerance when using the MoveIt Task Constructor—only when using the MoveGroupInterface class. I tried defining it in the ROS2 and MoveIt YAML files.

- [20240911T171410_motoros2_debug_log.txt](https://github.com/user-attachments/files/16970538/20240911T171410_motoros2_debug_log.txt)
- [PANELBOX.LOG](https://github.com/user-attachments/files/16970539/PANELBOX.LOG)

**Reply by gavanderhoorn** [reply, marked as answer]
According to your `PANELBOX.LOG` you have MotoROS2 `0.1.2` installed:

> ```
> //MOTOPLUS APP
> 01: MotoROS2 (humble)                0.1.2            micro-ROS based ROS 2 interface 
> ```

The fixes to goal tolerance parsing and checking are part of `0.1.3`. See the [CHANGELOG](https://github.com/Yaskawa-Global/motoros2/blob/7bac0079de3ce43288bc340e1d3b08a6e69d8fa2/CHANGELOG.md#013-2024-08-01).

Please download and install [MotoROS2 0.1.3](https://github.com/Yaskawa-Global/motoros2/releases/tag/0.1.3) and let us know whether that resolves your issue.

> I tried defining it in the ROS2 and MoveIt YAML files.

as I wrote above (https://github.com/Yaskawa-Global/motoros2/discussions/308#discussioncomment-10617705):

> A quick comment: `ros_control` is not used by MotoROS2, so changing `ros2_controllers.yaml` won't do anything. Neither the Moveit configuration files.

**Reply by SergeAlhalbi** [reply]
Hello, I downloaded and installed [MotoROS2 0.1.3](https://github.com/Yaskawa-Global/motoros2/releases/tag/0.1.3) and the error is now gone. Thank you very much for your help!

**Reply by gavanderhoorn** [reply]
Good to hear you got it to work.

**Reply by gavanderhoorn** [reply]
Note: the changes in `0.1.3` do *not* disable goal tolerance checks. They fix the checks on the MotoROS2 side. So if you provide 'incorrect' or impossible goal tolerances, the checks will still likely fail.

**Reply by SergeAlhalbi** [reply]
Thank you. I understand that. That's great!

**Comment by gavanderhoorn** [comment]
This appears to have been cross-posted to `stackoverflow`: [Final position was outside tolerance. Check robot safety limits that could be inhibiting motion](https://stackoverflow.com/questions/78975038). 

@SergeAlhalbi: I believe the moderators there would appreciate it if you could close your question while referencing the discussion here.

**Reply by SergeAlhalbi** [reply]
I posted an answer there and referenced our discussion. Thank you!
